api/crunch.ts

```diff
@@ -0,0 +1,224 @@
+    const result: GroupListItem[] = [];
+    for (let groupId in stats) {
+        const group = stats[groupId];
+        const groupItem: GroupListItem = { id: groupId, genres: [], all_tracks: [], members: Object.keys(group.members) };
+        result.push(groupItem);
+
+        for (let genreName in group.genres) {
+            const genre = group.genres[genreName];
+            const genreItem: GenreListItem = { name: genreName, tracks: [], score: 0 };
+            groupItem.genres.push(genreItem);
+
+            for (let trackId in genre.tracks) {
+                const track = genre.tracks[trackId];
+                const trackItem: ScoredTrack = { id: trackId, plays: track.plays, score: getTrackScore(group, track) };
+
+                genreItem.tracks.push(trackItem);
+                groupItem.all_tracks.push(trackItem);
+            }
+
+            sortByScore(genreItem.tracks);
+            genreItem.score = getGenreScore(groupItem, genreItem.tracks)
+        }
+
+        sortByScore(groupItem.all_tracks);
+    }
+
+    return result;
+}
+
+function getGenreScore(group: GroupListItem, tracks: ScoredTrack[]) {
+    var sum = tracks.reduce((sum, track) => sum + track.score, 0);
+    return sum / group.members.length;
+}
+
+function computePlaylists(data: GroupListItem[]) {
+    const statsNew: Playlist[] = [];
+    for (let group of data) {
+        statsNew.push({ groupId: group.id, name: "Top 10", score: 1, tracks: group.all_tracks.slice(0, 10) });
+
+        for (let genre of group.genres) {
+            statsNew.push({ groupId: group.id, name: genre.name, score: genre.score, tracks: genre.tracks.slice(0, 10) });
+        }
+    }
+    return statsNew;
+}
+
+function getTrackScore(group: Group, track: Track) {
+    return track.plays.map((trackPlays) => {
+        const totalPlays = group.members[trackPlays.user];
+        return trackPlays.plays / totalPlays;
+    }).reduce((sum, score) => sum + score, 0);
+}
+
+function sortByScore(tracks: ScoredTrack[]) {
+    return tracks.sort((a, b) => b.score - a.score);
+}
+
+async function createPlaylist(playlist: Playlist) {
+    await prisma.playlists.create({
+        data: {
+            name: playlist.name,
+            is_current_top_list: true,
+            score: playlist.score,
+            group_id: playlist.groupId,
+            playlist_items: {
+                createMany: {
+                    data: playlist.tracks.map((track) => {
+                        return {
+                            track_id: track.id,
+                            number_of_plays: track.plays.reduce((sum, userPlays) => sum + userPlays.plays, 0),
+                            number_of_unique_listeners: track.plays.length,
+                        }
+                    })
+                }
+            }
+        }
+    });
+}
+
+function fetchPlayedTracksLast7days(): Promise<PlayedTrack[]> {
+    return prisma.$queryRaw<PlayedTrack[]>`
+    SELECT 
```

Comment:
No way to achieve this without a raw query?

Reply:
Maybe, but I don't want to spend more time on wrestling with an ORM when I know how to do it without one :D